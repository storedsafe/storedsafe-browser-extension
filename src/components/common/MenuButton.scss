@import '../../variables';

.menu-button {
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  width: fit-content;

  .menu-button-title {
    position: absolute;
    display: flex;
    flex-direction: row;
    align-items: center;
    right: 0;
    background: $color-bg;
    color: $color-fg;
    margin-right: $menu-size + $spacing;
    height: $menu-size;
    width: max-content;
    border-radius: $border-radius;
    padding: $spacing;
    opacity: 0.8;
    visibility: hidden;
    transition-duration: 0s, $transition-duration, $transition-duration, $transition-duration;
    transition-delay: $transition-duration, 0s, 0s, 0s;
    transition-property: visibility, opacity, border-radius, margin-right;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  }

  .menu-button-icon:hover + .menu-button-title {
    transition-delay: 0s;
    visibility: visible;
    margin-right: $menu-size;
    opacity: 1;
    border-radius: $border-radius 0 0 $border-radius;
  }

  .menu-button-icon:focus + .menu-button-title {
    border: $border-width solid $color-accent-0;
    border-right-width: 0;
  }

  .menu-button-icon {
    border: 0;
    outline: 0;
    border: $border-width solid $color-bg-alt;
    background-color: $color-bg-alt;
    cursor: pointer;
    height: $menu-size;
    width: $menu-size;
    border-radius: $border-radius;
    padding: $spacing;
    transition: $transition-duration;
    transition-property: background-color, border-radius, border-color;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:hover {
      background-color: $color-bg;
      border-radius: 0 $border-radius $border-radius 0;
    }

    &.selected:not(:hover) {
      border-radius: 0 $border-radius $border-radius $border-radius;
    }

    > svg {
      height: 100%;
      width: 100%;
      transition: fill $transition-duration;
      fill: $color-primary-0;
    }

    &:focus {
      border-color: $color-accent-0;
    }

    &:focus:hover {
      border-left-color: transparent;
    }

    &:hover > svg,
    &:focus > svg,
    &.selected > svg {
      fill: $color-accent-0;
    }
  }
}
