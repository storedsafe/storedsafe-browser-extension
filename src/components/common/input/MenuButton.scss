@import '../../../styles/variables.scss';

.menu-button {
  border: 0;
  outline: 0;
  background-color: $color-primary-3;
  cursor: pointer;
  height: $menu-size;
  width: $menu-size;
  padding: $spacing;
  transition: background-color $transition-duration;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  border-radius: 0 $border-radius $border-radius $border-radius;
  box-sizing: border-box;

  > * {
    height: 100%;
    width: 100%;
  }

  > svg {
    fill: $color-fg-alt;
    transition: fill $transition-duration;
  }

  &:hover,
  &:focus {
    background-color: $color-bg;
  }

  &:hover > svg,
  &:focus > svg {
    fill: $color-primary-3;
  }
}
