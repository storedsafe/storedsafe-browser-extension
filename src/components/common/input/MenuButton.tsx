import React from 'react'
import './MenuButton.scss'

interface CustomMenuButtonProps {
  icon: React.ReactNode
}

type ButtonAttrs = React.ButtonHTMLAttributes<HTMLButtonElement>
type DetailedButtonAttrs = React.DetailedHTMLProps<ButtonAttrs, HTMLButtonElement>
type MenuButtonProps = CustomMenuButtonProps & DetailedButtonAttrs

export const MenuButton: React.FunctionComponent<MenuButtonProps> = ({
  icon,
  className,
  ...props
}: MenuButtonProps) => {
  const classNames = ['menu-button', className].join(' ')

  return (
    <button className={classNames} {...props}>
      {icon}
    </button>
  )
}
