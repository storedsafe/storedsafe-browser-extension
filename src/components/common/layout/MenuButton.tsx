import React from 'react'
import './MenuButton.scss'

interface MenuButtonProps {
  title: string
  icon: React.ReactNode
  onClick: () => void
  selected?: boolean
}

export const MenuButton: React.FunctionComponent<MenuButtonProps> = ({
  title,
  icon,
  onClick,
  selected
}: MenuButtonProps) => (
  <div className='menu-button' onClick={onClick}>
    <button
      className={`menu-button-icon${selected ? ' selected' : ''}`}
      aria-label={title}
    >
      {icon}
    </button>
    <div className='menu-button-title'>{title}</div>
  </div>
)

MenuButton.defaultProps = {
  selected: false
}
