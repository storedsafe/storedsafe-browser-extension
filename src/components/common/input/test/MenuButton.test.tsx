import * as React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { MenuButton } from '../MenuButton'
import pretty from 'pretty'

/**
 * Set up and tear down container to mount tested component in.
 * */
let container: Element = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})
afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

test('<MenuButton />', () => {
  act(() => {
    render(<MenuButton icon={<p>icon</p>} />, container)
  })

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
