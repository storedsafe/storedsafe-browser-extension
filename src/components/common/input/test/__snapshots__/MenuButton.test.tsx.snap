// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> 1`] = `
"<button class=\\"menu-button \\">
  <p>icon</p>
</button>"
`;
